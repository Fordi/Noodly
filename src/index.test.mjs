// eslint-disable-next-line
import { jest } from '@jest/globals';
import Noodly from './index.mjs';

describe('Noodly', () => {
  it('exists', () => {
    expect(Noodly instanceof Function).toBe(true);
  });
  it('creates a state machine', async () => {
    const store = Noodly({ value: 0 });
    const action = store.action({
      async* action() {
        yield { value: 1 };
        yield { value: 2 };
      },
    });
    expect(store.getState().value).toBe(0);
    await action();
    expect(store.getState().value).toBe(2);
  });
  it('allows signal subscription control', async () => {
    const store = Noodly({ value: 0 });
    const action = store.action({
      async* action() {
        yield { value: 1 };
        yield { value: 2 };
      },
    });
    const steps = [];
    const unlisten = store.listen((newState, oldState, signal, path) => {
      steps.push([{ ...newState }, { ...oldState }, signal, [...path]]);
    });
    await action();
    expect(steps).toMatchSnapshot();
    unlisten();
    steps.length = 0;
    await action();
    expect(steps.length).toBe(0);
  });
  it('talks to logs', async () => {
    const store = Noodly({ value: 0 });
    global.console.info = jest.fn();
    store.log(true);
    const action = store.action({
      async* action() {
        yield { value: 2 };
      },
    });
    await action();
    expect(global.console.info).toHaveBeenCalled();
  });
  it('rejects bad actions', () => {
    const store = Noodly({ value: 0 });
    expect(() => {
      store.action(() => {});
    }).toThrow();
  });
  it('supports child machines', async () => {
    const left = Noodly({ left: false });
    const leftAction = left.action({
      async* leftAction() {
        yield { hasRun: true };
      },
    });
    const right = Noodly({ right: false });
    const rightAction = right.action({
      async* rightAction() {
        yield { hasRun: true };
      },
    });
    const main = Noodly({}, { left, right });
    let steps = [];
    const unlisten = main.listen((newState, oldState, signal, path) => {
      steps.push([{ ...newState }, { ...oldState }, signal, [...path]]);
    });
    expect(main.getState()).toMatchSnapshot();
    await leftAction();
    await rightAction();
    expect(steps).toMatchSnapshot();
    unlisten();
    steps = [];
    await leftAction();
    await rightAction();
    expect(steps.length).toBe(0);
    expect(left.getState(1)).toBe(main.getState());
  });
  it('handle errors automatically, logging if logging is set', async () => {
    const store = Noodly({ });
    const action = store.action({
      // eslint-disable-next-line require-yield
      async* action() {
        throw new Error('failure');
      },
    });
    await action();
    expect(store.getState().error).toMatchSnapshot();
    store.log(true);
    global.console.warn = jest.fn();
    await action();
    expect(global.console.warn).toHaveBeenCalled();
  });
});
