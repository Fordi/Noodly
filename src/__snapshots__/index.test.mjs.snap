// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Noodly allows signal subscription control 1`] = `
Array [
  Array [
    Object {
      "value": 0,
    },
    Object {
      "value": 0,
    },
    "start",
    Array [],
  ],
  Array [
    Object {
      "value": 1,
    },
    Object {
      "value": 0,
    },
    "step",
    Array [],
  ],
  Array [
    Object {
      "value": 2,
    },
    Object {
      "value": 1,
    },
    "step",
    Array [],
  ],
  Array [
    Object {
      "value": 2,
    },
    Object {
      "value": 0,
    },
    "complete",
    Array [],
  ],
]
`;

exports[`Noodly handle errors automatically, logging if logging is set 1`] = `[Error: failure]`;

exports[`Noodly supports child machines 1`] = `
Object {
  "left": Object {
    "left": false,
  },
  "right": Object {
    "right": false,
  },
}
`;

exports[`Noodly supports child machines 2`] = `
Array [
  Array [
    Object {
      "left": Object {
        "left": false,
      },
      "right": Object {
        "right": false,
      },
    },
    Object {
      "left": Object {
        "left": false,
      },
      "right": Object {
        "right": false,
      },
    },
    "start",
    Array [
      "left",
    ],
  ],
  Array [
    Object {
      "left": Object {
        "hasRun": true,
      },
      "right": Object {
        "right": false,
      },
    },
    Object {
      "left": Object {
        "hasRun": true,
      },
      "right": Object {
        "right": false,
      },
    },
    "step",
    Array [
      "left",
    ],
  ],
  Array [
    Object {
      "left": Object {
        "hasRun": true,
      },
      "right": Object {
        "right": false,
      },
    },
    Object {
      "left": Object {
        "hasRun": true,
      },
      "right": Object {
        "right": false,
      },
    },
    "complete",
    Array [
      "left",
    ],
  ],
  Array [
    Object {
      "left": Object {
        "hasRun": true,
      },
      "right": Object {
        "right": false,
      },
    },
    Object {
      "left": Object {
        "hasRun": true,
      },
      "right": Object {
        "right": false,
      },
    },
    "start",
    Array [
      "right",
    ],
  ],
  Array [
    Object {
      "left": Object {
        "hasRun": true,
      },
      "right": Object {
        "hasRun": true,
      },
    },
    Object {
      "left": Object {
        "hasRun": true,
      },
      "right": Object {
        "hasRun": true,
      },
    },
    "step",
    Array [
      "right",
    ],
  ],
  Array [
    Object {
      "left": Object {
        "hasRun": true,
      },
      "right": Object {
        "hasRun": true,
      },
    },
    Object {
      "left": Object {
        "hasRun": true,
      },
      "right": Object {
        "hasRun": true,
      },
    },
    "complete",
    Array [
      "right",
    ],
  ],
]
`;
